lightning/src/ln/channelmanager.rs

```diff
@@ -333,29 +336,155 @@ mod inbound_payment {
-		let metadata_key = Hmac::from_engine(hmac).into_inner();
+/// LDK has multiple reasons to generate fake short channel ids:
+/// 1) zero-conf channels that don't have a confirmed channel id yet
+/// 2) phantom node payments, to get an scid for the phantom node's phantom channel
+mod fake_scid {
+	use bitcoin::blockdata::constants::genesis_block;
+	use bitcoin::hash_types::BlockHash;
+	use bitcoin::network::constants::Network;
+	use chain::keysinterface::{Sign, KeysInterface};
+	use ln::channelmanager::DecodeError;
+	use util::scid_utils;
+	use util::ser::Readable;
 
-		let mut hmac = HmacEngine::<Sha256>::new(&prk[..]);
-		hmac.input(&metadata_key);
-		hmac.input(&[2; 1]);
-		let ldk_pmt_hash_key = Hmac::from_engine(hmac).into_inner();
+	use core::convert::TryInto;
+	use core::ops::Deref;
 
-		let mut hmac = HmacEngine::<Sha256>::new(&prk[..]);
-		hmac.input(&ldk_pmt_hash_key);
-		hmac.input(&[3; 1]);
-		let user_pmt_hash_key = Hmac::from_engine(hmac).into_inner();
+	const TEST_SEGWIT_ACTIVATION_HEIGHT: u32 = 0;
+	const MAINNET_SEGWIT_ACTIVATION_HEIGHT: u32 = 481_824;
+	const MAX_TX_INDEX: u32 = 2_500;
+	const MAX_VOUT: u16 = 10_000;
+	const MAX_NAMESPACES: u8 = 8;
+	const NAMESPACE_ID_BITMASK: u8 = 0b111;
+
+	/// Fake scids are divided into namespaces, with each namespace having its own randomly-selected
+	/// identifier between [0..7], such that `scid_tx_index % MAX_NAMESPACES == namespace_identifier`.
+	/// This allows us to identify what namespace a fake scid corresponds to upon HTLC receipt, and
+	/// handle the HTLC accordingly.
+	pub(super) enum Namespace {
+		Phantom(u8),
+		// Coming soon: a variant for the zero-conf scid namespace
+	}
+
+	impl Namespace {
+		/// We generate "realistic-looking" random scids here, meaning the scid's block height is
+		/// between segwit activation and the current best known height, and the tx index and output
+		/// index are also selected from a "reasonable" range. We add this logic because it makes it
+		/// non-obvious at a glance that the scid is fake, e.g. if it appears in invoice route hints.
+		pub(super) fn get_fake_scid<Signer: Sign, K: Deref>(&self, highest_seen_blockheight: u32, genesis_hash: &BlockHash, keys_manager: &K) -> u64
+			where K::Target: KeysInterface<Signer = Signer>,
+		{
+			let scid_namespace = self.as_u8();
+			let rand_bytes = keys_manager.get_secure_random_bytes();
+
+			let valid_block_range = highest_seen_blockheight - segwit_activation_height(genesis_hash);
+			let rand_value_for_height = u32::from_be_bytes(rand_bytes[..4].try_into().unwrap());
+			let fake_scid_height = segwit_activation_height(genesis_hash) + rand_value_for_height % valid_block_range;
+
+			let rand_i32 = i32::from_be_bytes(rand_bytes[4..8].try_into().unwrap()).abs();
+			let rand_tx_index =  rand_i32 % (MAX_TX_INDEX as i32);
+			// Adding this offset to `rand_tx_index` will put it into the given `scid_namespace`.
+			let offset = (scid_namespace as i32 - rand_tx_index) % (MAX_NAMESPACES as i32);
```

Comment:
This code is so confusing can we just divide by 8, multiply by 8 and then add the scid_namespace?